[browser] Fix System.Delegate.DynamicInvoke method can not be resolved

kjpou1 opened the pull request:
- When building an app using `<PublishTrimmed>true</PublishTrimmed>` the linker is linking out System.Delegate.DynamicInvoke
- Call Delegate Invoke directly

---

Remove dependency on System.Delegate.DynamicInvoke
- Explicitly provide the declaring type to GetMethodFromHandle in GetCallSignature
- Fixes error resolving generic method handles.
   ```
      Error: System.ArgumentException: Cannot resolve method Int32 Invoke(Int32, Int32) because the declaring type of the method handle System.Func`3[T1,T2,TResult] is generic. Explicitly provide the declaring type to GetMethodFromHandle.
   ```

Add method to retrieve the Invoke method of the delegate

---

Added delegate tests. 

Fix https://github.com/dotnet/runtime/issues/47516

kjpou1 committed:
[browser] Fix System.Delegate.DynamicInvoke method can not be resolved

- When building an app using <PublishTrimmed>true</PublishTrimmed> the linker is linking out System.Delegate.DynamicInvoke
- Add this to the linker trim exclusions.

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/ILLink/ILLinkTrim.xml:
```diff
@@ -2,4 +2,12 @@
+    <!--
+    Preserve System.Delegate.DynamicInvoke which is called from javascript
+    -->
+    <assembly fullname="mscorlib">
```
It should be System.Private.CoreLib

kjpou1 replied:
hmmm that causes a build error.

```
resource System.Private.Runtime.InteropServices.JavaScript.xml in System.Private.Runtime.InteropServices.JavaScript, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a : warning IL2007: Could not resolve assembly 'System.Private.CoreLib' [/src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System.Private.Runtime.InteropServices.JavaScript.csproj]
    1 Warning(s)
    0 Error(s)
```

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/ILLink/ILLinkTrim.xml:
```diff
@@ -2,4 +2,12 @@
+    -->
+    <assembly fullname="mscorlib">
+      <type fullname="System.Delegate">
+        <method signature="System.Object DynamicInvoke(System.Object[])" />
```
Is it necessary to call via this slow path? Cannot we simply call Invoke directly or use even faster path?

@lewing

kjpou1 replied:
This is the only way to call this from JavaScript.  Look up the assembly and get a reference to the object of DynamicInvoke and then call the driver.

kjpou1 replied:
Will talk with Larry tonight

kjpou1 replied:
Took more of a look at this and we would need to add generics support to the bindings interface.

kjpou1 committed:
Address review comment for assembly name

kjpou1 committed:
Change this back to mscorlib as the recommended code cause a warning.

```
resource System.Private.Runtime.InteropServices.JavaScript.xml in System.Private.Runtime.InteropServices.JavaScript, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a : warning IL2007: Could not resolve assembly 'System.Private.CoreLib' [/src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System.Private.Runtime.InteropServices.JavaScript.csproj]
    1 Warning(s)
    0 Error(s)
```

eerhardt commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/ILLink/ILLinkTrim.xml:
```diff
@@ -2,4 +2,12 @@
         <type fullname="System.Runtime.InteropServices.JavaScript.Runtime" />
     </assembly>
+    <!--
+    Preserve System.Delegate.DynamicInvoke which is called from javascript
```
Is there any place in IL that can use `[DynamicDependency]` instead?

Another thought is to put a new method on `System.Runtime.InteropServices.JavaScript.Runtime` that JS invokes into to call the `Delegate.DynamicInvoke` method. The whole `Runtime` class is preserved, so nothing in it will be trimmed.

cc @vitek-karas

kjpou1 replied:
Thanks for the suggestions but this should all be resolved by calling Invoke on the Delegate as suggested above all.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-fix-delegate-trimming

kjpou1 committed:
[browser] Remove  linker trim exclusions for System.Delegate.DynamicInvoke

kjpou1 committed:
Remove dependency on System.Delegate.DynamicInvoke

- Explicitly provide the declaring type to GetMethodFromHandle in GetCallSignature
- Fixes error resolving generic method handles.
   ```
      Error: System.ArgumentException: Cannot resolve method Int32 Invoke(Int32, Int32) because the declaring type of the method handle System.Func`3[T1,T2,TResult] is generic. Explicitly provide the declaring type to GetMethodFromHandle.
   ```

kjpou1 committed:
Add method to retrieve the Invoke method of the delegate

kjpou1 committed:
Create another test module for Delegate tests

kjpou1 committed:
Add MarshalFunctionReturnAction that marshals an Action delegate from a Function

kjpou1 committed:
More tests, Actions and Delegate methods

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-fix-delegate-trimming

kjpou1 committed:
Add multi cast tests

kjpou1 committed:
Add anon and lamda del test

kjpou1 committed:
Add tests for marshaling TypedArray from different delegates.

radical commented on src/mono/wasm/runtime/binding_support.js:
```diff
@@ -1459,21 +1460,15 @@ var BindingSupportLib = {
+				if (typeof delegate_obj.__mono_delegate_invoke_sig__ === "undefined")
+					delegate_obj.__mono_delegate_invoke_sig__ = Module.mono_method_get_call_signature (delegate_obj.__mono_delegate_invoke__, delegateRoot.value);
+
+				return this.call_method (delegate_obj.__mono_delegate_invoke__, delegateRoot.value, delegate_obj.__mono_delegate_invoke_sig__, js_args);
```
Will `call_method` handle creating roots for the args? This removed `argsRoot.value = this.js_array_to_mono_array (js_args);`, so I'm wondering if that gets handled somewhere else now.

Also, if we are removing that, then `argsRoot` can be dropped completely.

kjpou1 replied:
Addressed in recent commit.  Removed the argsRoot as it is no longer being used.

DynamicInvoke used an array of objects, thus the rooted object.  The Invoke method call is taken care of in the normal process of the invoke.

kg replied:
call_method creates a result root but only for the duration of the call, you have to root it afterward.
args must be rooted explicitly because the actual arg list passed to the underlying method call impl is just some raw pointers and bits in the heap, but you don't need to necessarily root each arg if they're all in a managed array.

radical commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+        }
+
+        [Fact]
+        public static void MarshalAnonDelegateMethod()
```
These kinda tests could be combined like:
```csharp
        [Theory]
        [InlineData("CreateDelegateFromAnonymousMethod_VoidString")]
        [InlineData("CreateDelegateFromLambda_VoidString")]
        [InlineData("CreateDelegateFromMethod_VoidString")]
        [InlineData("CreateActionT_VoidString")]
        public static void InvokeDelegate_VoidString(string creator)
        {
            HelperMarshal._delegateCallResult = string.Empty;
            var s = Runtime.InvokeJS($@"
                var del = App.call_test_method (""{creator}"", [  ]);
                del(""Hic sunt dracones"");
            ");

            Assert.Equal("Notification received for: Hic sunt dracones", HelperMarshal._delegateCallResult);
        }
```

```csharp
        internal static string _delegateCallResult;
        private static Del CreateDelegateFromAnonymousMethod_VoidString()
        {
            // Instantiate the delegate.
            Del handler = delegate(string name) { _delegateCallResult = $"Notification received for: {name}"; };
            return handler;
        }

        private static Del CreateDelegateFromLambda_VoidString()
        {
            // Instantiate the delegate.
            Del handler = (string name) => { _delegateCallResult = $"Notification received for: {name}"; };
            return handler;
        }

        public static void DelegateMethod_VoidString(string name) => _delegateCallResult = $"Notification received for: {name}";

        private static Del CreateDelegateFromMethod_VoidString()
        {
            // Instantiate the delegate.
            Del handler = DelegateMethod_VoidString;
            return handler;
        }

        private static Action<string> CreateActionT_VoidString()
            => (string name) => _delegateCallResult = $"Notification received for: {name}";
```

kjpou1 replied:
Addressed in recent commits

radical commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+        }
+
+        [Fact]
+        public static void MarshalCustomMultiDelegateAcceptingString()
```
The multicast delegate ones can be combined too

kjpou1 replied:
addressed in recent commits

radical commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+
+        [Theory]
+        [MemberData(nameof(Object_Prototype))]
+        public static void MarshalFunctionAcceptingUint8Array(Function objectPrototype)
```
Should other argument types be tested too?

kjpou1 replied:
Will add more

kjpou1 replied:
Added in recent commits

radical commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+        }
+
+        [Fact]
+        public static void MarshalFunctionLooptyLoopIncrement()
```
What is it that we want to test here?

kjpou1 replied:
Calling the delegate in a loop with different values.

kjpou1 replied:
Marking this resolved as well.  Thought it was a valid test just to cover different tests.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-fix-delegate-trimming

kjpou1 committed:
Address review comment.  Remove unused variable argsroot.

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+        }
+
+        [Fact]
+        public static void MarshalFunction()
```
I guess it is the point of view.  I thought of it as testing the marshaling of the delegates but it does not matter really.  Will change those

radical replied:
that's fine. just a thought :)

kjpou1 replied:
Addressed in recent commits

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+            Runtime.InvokeJS(@"
+                var funcDelegate = App.call_test_method (""CreateFunctionDelegate"", [  ]);
+                var res = funcDelegate (10, 20);
+                App.call_test_method (""InvokeI32"", [ res, res ]);
```
An extra call to load up the return value that will be asserted.  Does it need to be taken out??

radical replied:
No, I was just curious if there was any other reason.

kjpou1 replied:
Can this be resolved??

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+            Runtime.InvokeJS(@"
+                var funcDelegate = App.call_test_method (""CreateFunctionDelegate"", [  ]);
+                var res = funcDelegate (10, 20);
+                for (x = 0; x < 1000; x++)
+                {
+                    res = funcDelegate (10, 20);
+                }
```
Calling the delegate in a loop.

radical replied:
Just curious.. what changes, or might change in a loop?

kjpou1 replied:
Thought it was a valid test just to cover different tests.  Should this be removed then??

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+        }
+
+        [Fact]
+        public static void MarshalActionFloatIntToIntInt()
```
Not sure really.  Are we testing marshaling here or the proper function of a delegate being marshaled and the invoke method called on it.  It seems different types of marshaling are being tested throughout the other tests.

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+        }
+
+        [Theory]
+        [MemberData(nameof(Object_Prototype))]
```
Is it really necessary here?

radical replied:
I meant, that AFAIU, this test is using that `Object_prototype` was just converting to string. So, I was asking if there are other cases that I'm missing, because of which we have it as a method arg.

kjpou1 replied:
Then we should add more tests.  Will add them

kjpou1 replied:
Added more tests in recent commits

kjpou1 committed:
Remove extra assert as per review

kjpou1 committed:
Address review of switched expected

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+            ");
+
+            Assert.Equal(50, HelperMarshal._funcActionBufferResultValue.Length);
+            Assert.Equal(HelperMarshal._funcActionBufferResultLengthValue, HelperMarshal._funcActionBufferResultValue.Length);
```
addressed in recent commits

kjpou1 committed:
As per review comment Rename all Marshal method names to Invoke

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+        }
+
+        [Fact]
+        public static void MarshalFunctionReturnAction()
```
addressed in recent commits

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs:
```diff
@@ -0,0 +1,214 @@
+            var clamped = new byte[10];
+            HelperMarshal._funcActionBufferResultValue = Uint8Array.From(clamped);
+            Assert.Equal(10, HelperMarshal._funcActionBufferResultValue.Length);
+            Assert.NotEqual("[object Uint8ClampedArray]", objectPrototype.Call(HelperMarshal._funcActionBufferResultValue));
```
addressed in recent commits

kjpou1 committed:
Add more array type tests for delegates.

kjpou1 committed:
Combine Delegate tests and Multi Cast delegate tests

kjpou1 commented:
@marek-safar @lewing @radical

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-fix-delegate-trimming

# Conflicts:
#	src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

kjpou1 committed:
Fix accessor after merge conflict
